File: Java/tophub.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

public class Tophub {

	private String baseurl = "https://api.tophubdata.com";
	private String apikey = "";

	public Tophub(String apikey) {
		this.apikey = apikey;
	}

	public Map<String, Object> nodes(int p) {
		return call("/nodes", Map.of("p", p));
	}

	public Map<String, Object> node(String hashid) {
		return call("/nodes/" + hashid, new HashMap<>());
	}

	public Map<String, Object> nodeHistorys(String hashid, String date) {
		return call("/nodes/" + hashid + "/historys", Map.of("date", date));
	}

	public Map<String, Object> search(String q, int p, String hashid) {
		return call("/search", Map.of("q", q, "p", p, "hashid", hashid));
	}

	public Map<String, Object> call(String endpoint, Map<String, Object> params) {
		try {
			String url = baseurl + endpoint;
			if (!params.isEmpty()) {
				url += "?" + getQueryString(params);
			}

			HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
			connection.setRequestMethod("GET");
			connection.setRequestProperty("Authorization", apikey);

			BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
			StringBuilder response = new StringBuilder();
			String line;

			while ((line = reader.readLine()) != null) {
				response.append(line);
			}

			reader.close();
			connection.disconnect();

			// Assuming the response is in JSON format
			// You may need to adjust this based on the actual response format
			// For simplicity, I'm returning a Map<String, Object>
			// You can create appropriate classes to represent the response structure
			return parseJson(response.toString());
		} catch (Exception e) {
			e.printStackTrace();
		}

		return null;
	}

	private String getQueryString(Map<String, Object> params) {
		StringBuilder queryString = new StringBuilder();
		for (Map.Entry<String, Object> entry : params.entrySet()) {
			if (queryString.length() > 0) {
				queryString.append("&");
			}
			queryString.append(entry.getKey()).append("=").append(entry.getValue());
		}
		return queryString.toString();
	}

	// You may use a JSON parsing library for a more robust solution
	private Map<String, Object> parseJson(String json) {
		// Implement your JSON parsing logic here
		// For simplicity, I'm returning null
		return null;
	}

	public static void main(String[] args) {
		// Example usage
		String apikey = "";
		Tophub tophub = new Tophub(apikey);
		Map<String, Object> resultNodes = tophub.nodes(1);
		Map<String, Object> resultNode = tophub.node("mproPpoq6O");
		Map<String, Object> resultNodeHistorys = tophub.nodeHistorys("mproPpoq6O", "2023-01-01");
		Map<String, Object> resultSearch = tophub.search("苹果", 1, "mproPpoq6O");
	}
}

File: Go/tophub.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
)

// Tophub represents the Tophub API client
type Tophub struct {
	baseURL string
	apikey  string
}

// NewTophub creates a new instance of the Tophub API client
func NewTophub(apikey string) *Tophub {
	return &Tophub{
		baseURL: "https://api.tophubdata.com",
		apikey:  apikey,
	}
}

// Nodes retrieves nodes from the Tophub API
func (t *Tophub) Nodes(p int) (map[string]interface{}, error) {
	return t.call("/nodes", map[string]interface{}{"p": p})
}

// Node retrieves a specific node from the Tophub API
func (t *Tophub) Node(hashid string) (map[string]interface{}, error) {
	return t.call("/nodes/"+hashid, map[string]interface{}{})
}

// NodeHistorys retrieves node history from the Tophub API
func (t *Tophub) NodeHistorys(hashid, date string) (map[string]interface{}, error) {
	return t.call("/nodes/"+hashid+"/historys", map[string]interface{}{"date": date})
}

// Search performs a search on the Tophub API
func (t *Tophub) Search(q string, p int, hashid string) (map[string]interface{}, error) {
	return t.call("/search", map[string]interface{}{"q": q, "p": p, "hashid": hashid})
}

// call is a generic function to make API calls
func (t *Tophub) call(endpoint string, params map[string]interface{}) (map[string]interface{}, error) {
	url := t.baseURL + endpoint
	if len(params) > 0 {
		queryString := url.Values{}
		for key, value := range params {
			queryString.Add(key, fmt.Sprintf("%v", value))
		}
		url += "?" + queryString.Encode()
	}

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", t.apikey)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	err = json.Unmarshal(body, &result)
	if err != nil {
		return nil, err
	}

	return result, nil
}

func main() {
	// Example usage
	apikey := ""
	tophub := NewTophub(apikey)
	resultNodes, _ := tophub.Nodes(1)
	resultNode, _ := tophub.Node("mproPpoq6O")
	resultNodeHistorys, _ := tophub.NodeHistorys("mproPpoq6O", "2023-01-01")
	resultSearch, _ := tophub.Search("苹果", 1, "mproPpoq6O")

	fmt.Println(resultNodes)
	fmt.Println(resultNode)
	fmt.Println(resultNodeHistorys)
	fmt.Println(resultSearch)
}
